text/0003-data-frames.md

```diff
@@ -0,0 +1,252 @@
+
+# Motivation
+
+[motivation]: #motivation
+
+The current system has a few unexpected limitations:
+
+- The top-level rows represent a table of rows, but it's unclear how to represent a top-level list of strings vs a stream of strings.
+- A similar ambiguity exists between an "object" (a data structure denoted by key/value pairs) and a table of one row
+- Inner-tables are modelled differently than top-level tables, leading to confusion
+- There is no way to currently represent a matrix
+- As rows are streamed instead of tables, it's not possible to predict how to display this information. This is currently mitigated by buffering some number of rows and treating them as one "table"
+- Likewise, since rows are streamed instead of tables, it's unclear what the user should expect if they request a column that is not present, as this column may appear in the following row instead.
+- When table data is serialized, there is a large amount of duplication, as columns are repeated with each row sent.
+- Additionally, there is currently no way to represent rows using a row literal. We propose a frame literal that will fill this role.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Data frame representation:
+
+```rust
+struct DataFrame {
+    headers: Option<Vec<String>>,
+    rows: Vec<Vec<Value>>,
+    partial_frame_id: Option<Uuid>,
+    is_object: bool,
+}
+```
+
+## Tables
+
+A self-contained table would look like this:
+
+```rust
+let frame = DataFrame {
+    headers: Some(vec!["name", "age"]),
+    rows: vec![
+        vec![Value::from("Bob"), Value::from(30)],
+        vec![Value::from("Sally"), Value::from(43)]
+    ],
+    partial_frame_id: None,
+    is_object: false,
+};
+```
+
+The above code could be created using this Nu syntax:
+
+```sh
+[name: [Bob, Sally], age: [30, 43]]
+```
+
+## Lists and matrices
+
+A self-contained list would look like this:
+
+```rust
+let list = DataFrame {
+    headers: None,
+    rows: vec![
+        vec![Value::from(1), Value::from(2), Value::from(3)]
+    ],
+    partial_frame_id: None,
+    is_object: false,
+}
+```
+
+The above code could be created using this Nu syntax:
+
+```sh
+[1, 2, 3]
+```
+
+## Objects (aka hash tables)
+
+A self-contained object would look like this:
+
+```rust
+let obj = DataFrame {
+    headers: Some(vec!["name", "level"]),
+    rows: vec![
+        vec![Value::from("Thomas"), Value::from(12)]
+    ],
+    partial_frame_id: None,
+    is_object: true,
+}
+```
+
+**Note:** we use the boolean in the table rather than enumeration because all processing on the frame remains uniform regardless of if the frame is a single row with headers vs an object. This simplifies algorithms to only have to work with the data directly, and we can later represent this data and/or serialize this data in a way that maintains the user's model.
+
+The above code could be created using this Nu syntax:
+
+```sh
+[name: Thomas, level: 12]
```

Comment:
If I have 100 rows of data, do I have to repeat the column names for each row? It may be nice to consider something like `[columns: [name, level], rows:[[Thomas, 12], [Fred, 15], [Mark, 3]]]`. Maybe not that exact syntax but you get my meaning.

Reply:
I give an example above for how to write a dataframe. This example is about "objects", or hash tables, so we only have one value per column.

Reply:
I think we must be talking passed each other because I really understand what you're saying and I think you didn't understand what I was saying. I'm just showing a possible way of creating a dataframe literal without repeating the column names for every row. I define the column names one time with `[columns: [name, level]` and then add the rows with `[rows:[[Thomas, 12], [Fred, 15], [Mark, 3]]]`. `[Thomas, 12]` is one row, `[Fred, 15]` is another row, and `[Mark, 3]` is the last row.